### ISSUE TITLE
[bug] [jruby] sequential HTML "void" elements are nesting within each other on JRuby

### ISSUE THREAD

**Issue by postmodern** [sparklemotion/nokogiri#2141]
## Description

It appears that sequential HTML void elements (ex: `<iframe ... />`) are nesting within each other, instead of remaining sequential sibling elements, but only on JRuby.

## Steps to Reproduce

```
require 'nokogiri'
html = %{<html><body><iframe src="/one" /><iframe src="/two" /></body></html>}
doc = Nokogiri::HTML::Document.parse(html)
pp doc
```

## CRuby Output

```
#(Document:0x140 {
  name = "document",
  children = [
    #(DTD:0xb4 { name = "html" }),
    #(Element:0x12c {
      name = "html",
      children = [
        #(Element:0x118 {
          name = "body",
          children = [
            #(Element:0xdc {
              name = "iframe",
              attributes = [ #(Attr:0xc8 { name = "src", value = "/one" })]
              }),
            #(Element:0x104 {
              name = "iframe",
              attributes = [ #(Attr:0xf0 { name = "src", value = "/two" })]
              })]
          })]
      })]
  })
```

### JRuby Output

```
#(Document:0x7de {
  name = "document",
  children = [
    #(Element:0x7dc {
      name = "html",
      children = [
        #(Element:0x7d2 { name = "head" }),
        #(Element:0x7da {
          name = "body",
          children = [
            #(Element:0x7d8 {
              name = "iframe",
              attributes = [ #(Attr:0x7d4 { name = "src", value = "/one" })],
              children = [ #(Text "<iframe src=\"/two\" /></body></html>")]
              })]
          })]
      })]
  })
```

Note that the second `iframe` is now within the children of the first `iframe`.

## Expected Output

For the JRuby document tree to match that of CRuby's.

### CRuby

ruby 2.7.2p137 (2020-10-01 revision 5445e04352) [x86_64-linux]

#### Nokogiri (1.10.10)
    ---
    warnings: []
    nokogiri: 1.10.10
    ruby:
      version: 2.7.2
      platform: x86_64-linux
      description: ruby 2.7.2p137 (2020-10-01 revision 5445e04352) [x86_64-linux]
      engine: ruby
    libxml:
      binding: extension
      source: packaged
      libxml2_path: "/home/postmodern/.gem/ruby/2.7.2/gems/nokogiri-1.10.10/ports/x86_64-pc-linux-gnu/libxml2/2.9.10"
      libxslt_path: "/home/postmodern/.gem/ruby/2.7.2/gems/nokogiri-1.10.10/ports/x86_64-pc-linux-gnu/libxslt/1.1.34"
      libxml2_patches:
      - 0001-Revert-Do-not-URI-escape-in-server-side-includes.patch
      - 0002-Remove-script-macro-support.patch
      - 0003-Update-entities-to-remove-handling-of-ssi.patch
      - 0004-libxml2.la-is-in-top_builddir.patch
      - 0005-Fix-infinite-loop-in-xmlStringLenDecodeEntities.patch
      libxslt_patches: []
      compiled: 2.9.10
      loaded: 2.9.10

### JRuby

jruby 9.2.14.0 (2.5.7) 2020-12-08 ebe64bafb9 OpenJDK 64-Bit Server VM 25.275-b01 on 1.8.0_275-b01 +jit [linux-x86_64]

#### Nokogiri (1.10.10)
    ---
    warnings: []
    nokogiri: 1.10.10
    ruby:
      version: 2.5.7
      platform: java
      description: jruby 9.2.14.0 (2.5.7) 2020-12-08 ebe64bafb9 OpenJDK 64-Bit Server
        VM 25.275-b01 on 1.8.0_275-b01 +jit [linux-x86_64]
      engine: jruby
      jruby: 9.2.14.0
    xerces: Xerces-J 2.12.0
    nekohtml: NekoHTML 1.9.21

## Test

```
#!/usr/bin/env ruby

require 'nokogiri'
require 'minitest/autorun'

class Test < MiniTest::Spec
  describe "Nokogiri::HTML::Document.parse" do
    it "should not implicitly nest void elements within each other" do
      html = %{<html><body><iframe src="/one" /><iframe src="/two" /></body></html>}
      doc = Nokogiri::HTML::Document.parse(html)

      assert(doc.at_xpath('/html/body/iframe[@src="/one"]') != nil, "could not find the first iframe")
      assert(doc.at_xpath('/html/body/iframe[@src="/two"]') != nil, "could not find the second iframe")
    end
  end
end
```

**Comment by flavorjones** [comment]
@postmodern Thanks for submitting this issue! Sorry you're experiencing inconsistencies between the CRuby/libxml2 and JRuby/NekoHTML parsers.

Nokogiri tries to be a thin wrapper -- as thin as reasonable -- around the underlying parsers. What you're seeing is a difference in how those parsers recover from parsing errors when they see ill-formed markup; and so there's nothing Nokogiri can easily do to change that behavior.

That said, looking into this particular issue ... the first thing I did was to add `puts doc.errors` after the document parse call, and on JRuby I see:

```
Element <iframe> not closed properly.
Element <body> not closed properly.
Element <html> not closed properly.
```

Looking at the markup, I see that you're using self-closing tags for the `iframe` elements. If I modify the test to be:

```
    it "should not implicitly nest void elements within each other" do
      html = %{<html><body><iframe src="/one" ></iframe><iframe src="/two" ></iframe></body></html>}
      doc = Nokogiri::HTML::Document.parse(html)
      puts doc.errors.inspect
      assert(doc.at_xpath('/html/body/iframe[@src="/one"]') != nil, "could not find the first iframe")
      assert(doc.at_xpath('/html/body/iframe[@src="/two"]') != nil, "could not find the second iframe")
    end
```

then the errors go away, and the test passes.

My conclusion here is that libxml2's parser recovers from seeing a self-closing tag by inserting the closing tag into the DOM; and that NekoHTML's parser recovers by waiting until it sees a tag that cannot be enclosed before inserting the closing tag.

In conclusion: apologies that the underlying parsers are recovering differently; and apologies that Nokogiri cannot easily change this behavior. If you can, including the explicit end tags (rather than using self-closing tags) is a way to work around this problem.

Unfortunately, there's lots of ill-formed markup on the internet and being unable to consistently parse it all is frustrating. Sorry I can't be of more help here.
